datafusion/functions-aggregate/src/array_agg.rs

```diff
@@ -994,6 +1002,34 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn does_not_over_account_memory_for_merge() -> Result<()> {
+        let (mut acc1, mut acc2) = ArrayAggAccumulatorBuilder::string().build_two()?;
+
+        let a1 = ListArray::from_iter_primitive::<UInt64Type, _, _>(vec![
+            Some(vec![Some(0), Some(1), Some(2)]),
+            Some(vec![Some(3)]),
+            None,
+            Some(vec![Some(4)]),
+        ]);
+        let a2 = ListArray::from_iter_primitive::<UInt64Type, _, _>(vec![
+            Some(vec![Some(0), Some(1), Some(2)]),
+            Some(vec![Some(3)]),
+            None,
+            Some(vec![Some(4)]),
+        ]);
+
+        acc1.merge_batch(&[Arc::new(a1.slice(0, 1))])?;
+        acc2.merge_batch(&[Arc::new(a2.slice(0, 1))])?;
+
+        acc1 = merge(acc1, acc2)?;
```

Comment:
The `merge_batch` functions do not receive arbitrary data, they receive the results of calling `state()` in other accumulators. A fairer test would be to do something like:

```suggestion
        acc1.update_batch(&[Arc::clone(a1.values())])?;
        acc2.update_batch(&[Arc::clone(a2.values())])?;

        acc1 = merge(acc1, acc2)?;
```

If you run this, you would notice that the test actually passes without your changes.